ml-agents/mlagents/trainers/sac/policy.py

```diff
@@ -0,0 +1,318 @@
+        :param feed_dict: Feed dict needed update
+        :param update_dit: Update dict that needs update
+        :param stats_needed: Stats needed to get from the update.
+        :param reward_signal_minibatches: Minibatches to use for updating the reward signals,
+            indexed by name.
+        """
+        for name, r_mini_batch in reward_signal_minibatches.items():
+            feed_dict.update(
+                self.reward_signals[name].prepare_update(
+                    self.model, r_mini_batch, num_sequences
+                )
+            )
+            update_dict.update(self.reward_signals[name].update_dict)
+            stats_needed.update(self.reward_signals[name].stats_name_to_update_name)
+
+    def construct_feed_dict(
+        self, model: SACModel, mini_batch: Dict[str, Any], num_sequences: int
+    ) -> Dict[tf.Tensor, Any]:
+        """
+        Builds the feed dict for updating the SAC model.
+        :param model: The model to update. May be different when, e.g. using multi-GPU.
+        :param mini_batch: Mini-batch to use to update.
+        :param num_sequences: Number of LSTM sequences in mini_batch.
+        """
+        feed_dict = {
+            self.model.batch_size: num_sequences,
+            self.model.sequence_length: self.sequence_length,
+            self.model.next_sequence_length: self.sequence_length,
+            self.model.mask_input: mini_batch["masks"],
+        }
+        for name in self.reward_signals:
+            feed_dict[model.rewards_holders[name]] = mini_batch[
+                "{}_rewards".format(name)
+            ]
+
+        if self.use_continuous_act:
+            feed_dict[model.action_holder] = mini_batch["actions"]
+        else:
+            feed_dict[model.action_holder] = mini_batch["actions"]
+            if self.use_recurrent:
+                feed_dict[model.prev_action] = mini_batch["prev_action"]
+            feed_dict[model.action_masks] = mini_batch["action_mask"]
+        if self.use_vec_obs:
+            feed_dict[model.vector_in] = mini_batch["vector_obs"]
+            feed_dict[model.next_vector_in] = mini_batch["next_vector_in"]
+        if self.model.vis_obs_size > 0:
+            for i, _ in enumerate(model.visual_in):
+                _obs = mini_batch["visual_obs%d" % i]
+                feed_dict[model.visual_in[i]] = _obs
+            for i, _ in enumerate(model.next_visual_in):
+                _obs = mini_batch["next_visual_obs%d" % i]
+                feed_dict[model.next_visual_in[i]] = _obs
+        if self.use_recurrent:
+            mem_in = [
+                mini_batch["memory"][i]
+                for i in range(0, len(mini_batch["memory"]), self.sequence_length)
+            ]
+            # LSTM shouldn't have sequence length <1, but stop it from going out of the index if true.
+            offset = 1 if self.sequence_length > 1 else 0
+            next_mem_in = [
+                mini_batch["memory"][i][: self.m_size // 4]
```

Comment:
Why is this `// 4`?  What does the 4 represent?

Reply:
It's because this is used to populate the target network, which only has one of the 4 possible LSTM heads (policy network has Value, Q1, Q2, Policy and target only has Value). The value is always the last in the memory.

Reply:
This makes sense, but it's also unintuitive -- no need to change at this time but might want to think how we can better represent multiple lstms

Reply:
Ideally we have a memory in for each of the heads, I guess. That would be more intuitive. But this does add to the number of fields we're shuttling around in Barracuda and elsewhere